### ISSUE TITLE
A typemap default code variable would be useful

### ISSUE THREAD

**Issue by dannyvalentesonos** [swig/swig#2521]
Sometimes I need to declare a `directorin` typemap just so that I can tell it what java descriptor to use, using the `descriptor` attribute. For example:

```
%typemap(directorin, descriptor="Ljava/nio/ByteBuffer;", noblock=1) ScopedDirectByteBuffer {
}
```

However, doing this forces me to have to write out all the code that would normally be generated automatically by Swig for that typemap. I only want to set the `descriptor` attribute, but keep the default code generation. It would be nice to define a typemap without given any contents, or if contents are needed, a $default variable would be really nice.

**Comment by ojwb** [comment]
`"$typemap(directorin, ScopedDirectByteBuffer)"` as the body?  That might result in a typemap expansion loop though.  You didn't provide a complete example so I can't easily test.

**Comment by dannyvalentesonos** [comment]
Yes I've actually done exactly that @ojwb  but it fails with a warning that that would cause a loop. So that doesn't work.

**Comment by wsfulton** [comment]
The `descriptor` attribute being part of the 'directorin' typemap is arguably a design mistake for the Java's director typemaps. If this was done again, I'd make it a separate typemap, but now we are kind of stuck with it. Anyway, what you want is achievable via a copy and specialize approach. 

First make a [typemap copy](https://swig.org/Doc4.2/Typemaps.html#Typemaps_nn13), which copies the attributes and content to create a named typemap. Then override/specialize just the attributes and copy the content from the named copy via the [`$typemap` special variable macro](https://swig.org/Doc4.2/Typemaps.html#Typemaps_special_macro_typemap) as follows:
```swig
// Create a named typemap which is just a copy of the default SWIGTYPE typemap
%typemap(directorin) ScopedDirectByteBuffer GENERIC = SWIGTYPE;

// 'Specialize' the attributes for all ScopedDirectByteBuffer instances passed by value and copy content from the named typemap
%typemap(directorin, descriptor="Ljava/nio/ByteBuffer;", noblock=1) ScopedDirectByteBuffer {
  $typemap(directorin, ScopedDirectByteBuffer GENERIC)
}
```

**Comment by dannyvalentesonos** [comment]
> The `descriptor` attribute being part of the 'directorin' typemap is arguably a design mistake for the Java's director typemaps. If this was done again, I'd make it a separate typemap, but now we are kind of stuck with it. Anyway, what you want is achievable via a copy and specialize approach.
> 
> First make a [typemap copy](https://swig.org/Doc4.2/Typemaps.html#Typemaps_nn13), which copies the attributes and content to create a named typemap. Then override/specialize just the attributes and copy the content from the named copy via the [`$typemap` special variable macro](https://swig.org/Doc4.2/Typemaps.html#Typemaps_special_macro_typemap) as follows:
> 
> ```c++
> // Create a named typemap which is just a copy of the default SWIGTYPE typemap
> %typemap(directorin) ScopedDirectByteBuffer GENERIC = SWIGTYPE;
> 
> // 'Specialize' the attributes for all ScopedDirectByteBuffer instances passed by value and copy content from the named typemap
> %typemap(directorin, descriptor="Ljava/nio/ByteBuffer;", noblock=1) ScopedDirectByteBuffer {
>   $typemap(directorin, ScopedDirectByteBuffer GENERIC)
> }
> ```

I will try this. Thank you for the suggestion!

**Comment by ojwb** [comment]
> what you want is achievable via a copy and specialize approach.

Can you document this as a solution in the manual, and then we can close this?

It may seem obvious once you know, but it didn't occur to me before when I've hit situations myself where I wanted to use the existing typemap but do something extra before and/or after it.  We've had other reports from users wanting to do similar things too.

**Comment by wsfulton** [comment]
Hopefully c1cc807 helps.
